### PR TITLE
Added a default stacked layout for tab content.

### PR DESCRIPTION
**Pull request by yanniboi**

### CODE REVIEW

**Comment by yanniboi** [comment]
This is going on top of #63

File: contacts.module [hunk 1]
```diff
@@ -85,7 +86,40 @@ function template_preprocess_contacts_dash_tabs(array &$variables) {
  *   - region_attributes: HTML markup attributes for the content wrapper.
  */
 function template_preprocess_contacts_dash_tab_content(array &$variables) {
+  $variables['messages'] = $variables['content']['messages'];
   $variables['region_attributes'] = new Attribute($variables['region_attributes']);
+  /* @var \Drupal\contacts\Entity\ContactTab $tab */
+  $tab = \Drupal::service('contacts.tab_manager')->getTabByPath($variables['user'], $variables['subpage']);
+  $layout = $tab->get('layout') ?: 'contacts_tab_content.stacked';
+
+  // @todo switch to core layout manager plugin.manager.core.layout?
```
Comment (andrewbelcher) [inline contacts.module]: Yes - definitely want to use the core layout module, including upping our core version requirement. No point implement the non core stuff...

File: contacts.module [hunk 2]
```diff
@@ -85,7 +86,40 @@ function template_preprocess_contacts_dash_tabs(array &$variables) {
...
+  /* @var \Drupal\layout_plugin\Plugin\Layout\LayoutPluginManager $layoutPluginManager */
+  $layoutPluginManager = \Drupal::service('plugin.manager.layout_plugin');
+  /* @var \Drupal\layout_plugin\Plugin\Layout\LayoutInterface $layoutInstance */
+  $layoutInstance = $layoutPluginManager->createInstance($layout, []);
+
+  $regions = [];
+
+  foreach ($layoutInstance->getPluginDefinition()['regions'] as $key => $region) {
+    if (!empty($variables['content'][$key])) {
+      $regions[$key] = $variables['content'][$key];
```
Comment (andrewbelcher) [inline contacts.module]: Mentioned in #63, but we should definitely move building the block render array into here and probably move it into a render element for OO bliss...

File: contacts.module [hunk 3]
```diff
@@ -85,7 +86,40 @@ function template_preprocess_contacts_dash_tabs(array &$variables) {
  *   - region_attributes: HTML markup attributes for the content wrapper.
  */
 function template_preprocess_contacts_dash_tab_content(array &$variables) {
+  $variables['messages'] = $variables['content']['messages'];
```
Comment (andrewbelcher) [inline contacts.module]: What is done with `messages`?

File: src/Controller/DashboardController.php [hunk 4]
```diff
@@ -32,18 +33,28 @@ class DashboardController extends ControllerBase {
...
+   * @param \Drupal\layout_plugin\Plugin\Layout\LayoutPluginManager $layout_manager
+   *   The layout plugin manager.
    *
    * @todo Switch to core layout manager.
    */
-  public function __construct(ContactsTabManager $tab_manager, BlockManager $block_manager) {
+  public function __construct(ContactsTabManager $tab_manager, BlockManager $block_manager, LayoutPluginManager $layout_manager) {
     $this->tabManager = $tab_manager;
     $this->blockManager = $block_manager;
+    $this->layoutManager = $layout_manager;
```
Comment (andrewbelcher) [inline src/Controller/DashboardController.php]: Moving to a render element should remove the need for this!

File: src/Plugin/Block/ContactsDashboardTabs.php [hunk 5]
```diff
@@ -59,10 +67,15 @@ class ContactsDashboardTabs extends BlockBase implements ContextAwarePluginInter
...
+   * @param \Drupal\layout_plugin\Plugin\Layout\LayoutPluginManager $layout_manager
+   *   The layout plugin manager.
+   *
+   * @todo Switch to core layout manager.
    */
-  public function __construct(array $configuration, $plugin_id, $plugin_definition, ContactsTabManager $tab_manager) {
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, ContactsTabManager $tab_manager, LayoutPluginManager $layout_manager) {
     parent::__construct($configuration, $plugin_id, $plugin_definition);
     $this->tabManager = $tab_manager;
+    $this->layoutManager = $layout_manager;
```
Comment (andrewbelcher) [inline src/Plugin/Block/ContactsDashboardTabs.php]: Another duplication a render element will solve...

File: templates/contacts-dash-tab-content.html.twig [hunk 6]
```diff
@@ -1,8 +1,3 @@
 <div {{ attributes.addClass('dash-content') }}>
-  <div class="dash-left">
-    {{ content.left }}
-  </div>
-  <div class="dash-right">
-    {{ content.right }}
-  </div>
+  {{ content }}
```
Comment (andrewbelcher) [inline templates/contacts-dash-tab-content.html.twig]: Do we even need this now we've got layout handling it?

**Review by andrewbelcher — CHANGES_REQUESTED** [review]
Great direction. Definitely want to switch to core layout and move the layout and block rendering into a render element.

Also - couldn't comment directly on the image, but it's 2 col, but not stacked!

File: contacts.module [hunk 7]
```diff
@@ -75,17 +66,15 @@ function template_preprocess_contacts_dash_tabs(array &$variables) {
...
  *
- * @param array $variables
- *   An associative array containing:
- *   - attributes: HTML markup attributes for the content wrapper.
- *   - region_attributes: HTML markup attributes for the content wrapper.
+ * @see template_preprocess_contacts_dash_tab_content()
  */
-function template_preprocess_contact_tab_content(array &$variables) {
-  $variables['region_attributes'] = new Attribute($variables['region_attributes']);
+function template_preprocess_contacts_tab_content_stacked(array &$variables) {
```
Comment (andrewbelcher) [inline contacts.module]: Let's drop the pre-process, we can use `$content['region']['#prefix']` etc to add additional wrappers as required for manage mode.

File: layouts/tab_content_stacked/contacts-tab-content-stacked.html.twig [hunk 8]
```diff
@@ -0,0 +1,14 @@
+<div {{ attributes }}>
+  <div {{ region_attributes.addClass('dash-top') }}>
```
Comment (andrewbelcher) [inline layouts/tab_content_stacked/contacts-tab-content-stacked.html.twig]: Apply the classes directly rather than via `region_attributes`.

**Review by andrewbelcher — CHANGES_REQUESTED** [review]

**Pull request merged**